File: src/algorithms/python/astar.py
import heapq

def heuristic(a, b):
    """
    Calculate Manhattan distance heuristic between two points.
    
    Args:
        a (tuple): First point (row, col)
        b (tuple): Second point (row, col)
        
    Returns:
        int: Manhattan distance
    """
    return abs(a[0] - b[0]) + abs(a[1] - b[1])


def astar(grid, start, end):
    """
    A* Search Algorithm for Pathfinding
    Time Complexity: O(b^d) where b is branching factor and d is depth
    Space Complexity: O(b^d)
    
    A* is an informed search algorithm that uses heuristics to guide its search.
    It combines the benefits of Dijkstra's algorithm and greedy best-first search.
    
    The algorithm uses:
    - g(n): Cost from start to current node
    - h(n): Estimated cost from current node to end (heuristic)
    - f(n): Total estimated cost = g(n) + h(n)
    
    Args:
        grid (list): 2D grid representing the map (0 = walkable, 1 = wall)
        start (tuple): Starting position (row, col)
        end (tuple): Target position (row, col)
        
    Returns:
        list: Path from start to end as list of (row, col) tuples, or None if no path exists
    """
    rows, cols = len(grid), len(grid[0])
    
    # Priority queue: stores (f_score, g_score, row, col)
    # f_score is used for priority, g_score is used as tiebreaker
    pq = [(0, 0, start[0], start[1])]
    
    # Dictionary to store g_score (cost from start) for each cell
    g_scores = {start: 0}
    
    # Dictionary to store f_score (g + h) for each cell
    f_scores = {start: heuristic(start, end)}
    
    # Dictionary to reconstruct the path
    came_from = {}
    
    # Set to track cells in the open set
    open_set = {start}
    
    # Directions: up, down, left, right
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    while pq:
        current_f, current_g, row, col = heapq.heappop(pq)
        current = (row, col)
        
        # Remove from open set
        open_set.discard(current)
        
        # If we reached the end, reconstruct the path
        if current == end:
            path = []
            while current in came_from:
                path.append(current)
                current = came_from[current]
            path.append(start)
            return path[::-1]  # Reverse to get start -> end
        
        # Skip if we've already found a better path to this cell
        if current_g > g_scores.get(current, float('inf')):
            continue
        
        # Explore all 4 directions
        for dr, dc in directions:
            new_row, new_col = row + dr, col + dc
            neighbor = (new_row, new_col)
            
            # Check if the new position is valid
            if (0 <= new_row < rows and 
                0 <= new_col < cols and 
                grid[new_row][new_col] == 0):
                
                # Calculate tentative g_score (cost from start to neighbor)
                tentative_g = current_g + 1
                
                # If we found a better path to this neighbor
                if tentative_g < g_scores.get(neighbor, float('inf')):
                    # Update path
                    came_from[neighbor] = current
                    g_scores[neighbor] = tentative_g
                    
                    # Calculate f_score = g + h
                    h_score = heuristic(neighbor, end)
                    f_score = tentative_g + h_score
                    f_scores[neighbor] = f_score
                    
                    # Add to open set if not already there
                    if neighbor not in open_set:
                        open_set.add(neighbor)
                        heapq.heappush(pq, (f_score, tentative_g, new_row, new_col))
    
    # No path found
    return None


# Example usage and test
if __name__ == "__main__":
    # Create a sample grid (0 = walkable, 1 = wall)
    test_grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 1, 1, 0],
        [0, 0, 0, 0, 0]
    ]
    
    start_pos = (0, 0)
    end_pos = (4, 4)
    
    print(f"Grid size: {len(test_grid)}x{len(test_grid[0])}")
    print(f"Start: {start_pos}, End: {end_pos}")
    print(f"Heuristic (Manhattan distance): {heuristic(start_pos, end_pos)}")
    
    path = astar(test_grid, start_pos, end_pos)
    
    if path:
        print(f"\nPath found with {len(path)} steps:")
        print(f"Path: {path}")
        print(f"Actual distance: {len(path) - 1}")
    else:
        print("No path found!")
    
    # Visualize the path on the grid
    if path:
        print("\nGrid with path (. = walkable, # = wall, * = path):")
        for i in range(len(test_grid)):
            row_str = ""
            for j in range(len(test_grid[0])):
                if (i, j) in path:
                    row_str += "* "
                elif test_grid[i][j] == 1:
                    row_str += "# "
                else:
                    row_str += ". "
            print(row_str)

File: src/algorithms/pathfinding/aStar.js
import { GRID_ELEMENT_STATES } from '../../constants';

/**
 * Manhattan distance heuristic for A*
 * @param {Object} a - Position {row, col}
 * @param {Object} b - Position {row, col}
 * @returns {number} - Manhattan distance
 */
function manhattanDistance(a, b) {
  return Math.abs(a.row - b.row) + Math.abs(a.col - b.col);
}

/**
 * A* Pathfinding Algorithm
 * Time Complexity: O(b^d) where b is branching factor and d is depth
 * Space Complexity: O(b^d)
 *
 * A* uses heuristics to guide its search, combining benefits of Dijkstra and greedy best-first search.
 * Uses f(n) = g(n) + h(n) where g(n) is cost from start and h(n) is heuristic to goal.
 *
 * @param {number[][]} grid - 2D grid with cell weights
 * @param {Object} start - Start position {row, col}
 * @param {Object} end - End position {row, col}
 * @param {number} rows - Number of rows in grid
 * @param {number} cols - Number of columns in grid
 * @returns {Object[]} - Array of animation steps
 */
export function aStar(grid, start, end, rows, cols) {
  const steps = [];

  if (!start || !end || !rows || !cols) {
    console.error('A*: Invalid inputs', { start, end, rows, cols });
    return steps;
  }

  if (
    start.row < 0 ||
    start.row >= rows ||
    start.col < 0 ||
    start.col >= cols
  ) {
    console.error('A*: Start position out of bounds', { start, rows, cols });
    return steps;
  }

  if (end.row < 0 || end.row >= rows || end.col < 0 || end.col >= cols) {
    console.error('A*: End position out of bounds', { end, rows, cols });
    return steps;
  }

  const gScore = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(Infinity));
  const fScore = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(Infinity));
  const visited = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(false));
  const parent = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(null));

  const states = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(GRID_ELEMENT_STATES.DEFAULT));
  states[start.row][start.col] = GRID_ELEMENT_STATES.START;
  states[end.row][end.col] = GRID_ELEMENT_STATES.END;

  steps.push({
    grid: grid.map(row => [...row]),
    states: states.map(row => [...row]),
    description: 'Starting A* pathfinding',
  });

  gScore[start.row][start.col] = 0;
  fScore[start.row][start.col] = manhattanDistance(start, end);

  const openSet = [
    {
      row: start.row,
      col: start.col,
      f: fScore[start.row][start.col],
    },
  ];

  const directions = [
    { row: -1, col: 0 }, // up
    { row: 1, col: 0 }, // down
    { row: 0, col: -1 }, // left
    { row: 0, col: 1 }, // right
  ];

  let found = false;

  while (openSet.length > 0 && !found) {
    openSet.sort((a, b) => a.f - b.f);
    const current = openSet.shift();

    if (visited[current.row][current.col]) {
      continue;
    }

    visited[current.row][current.col] = true;

    if (
      !(current.row === start.row && current.col === start.col) &&
      !(current.row === end.row && current.col === end.col)
    ) {
      states[current.row][current.col] = GRID_ELEMENT_STATES.CLOSED;
    }

    const g = gScore[current.row][current.col];
    const h = manhattanDistance(current, end);

    steps.push({
      grid: grid.map(row => [...row]),
      states: states.map(row => [...row]),
      description: `Exploring (${current.row}, ${current.col}) | g=${g.toFixed(1)}, h=${h.toFixed(1)}, f=${current.f.toFixed(1)}`,
    });

    if (current.row === end.row && current.col === end.col) {
      found = true;
      break;
    }

    for (const dir of directions) {
      const newRow = current.row + dir.row;
      const newCol = current.col + dir.col;

      if (newRow < 0 || newRow >= rows || newCol < 0 || newCol >= cols) {
        continue;
      }

      if (visited[newRow][newCol]) {
        continue;
      }

      const tentativeG = gScore[current.row][current.col] + 1;

      if (tentativeG < gScore[newRow][newCol]) {
        parent[newRow][newCol] = current;
        gScore[newRow][newCol] = tentativeG;
        fScore[newRow][newCol] =
          tentativeG + manhattanDistance({ row: newRow, col: newCol }, end);

        openSet.push({
          row: newRow,
          col: newCol,
          f: fScore[newRow][newCol],
        });

        if (!(newRow === end.row && newCol === end.col)) {
          states[newRow][newCol] = GRID_ELEMENT_STATES.OPEN;
        }
      }
    }
  }

  if (found) {
    const path = [];
    let current = end;

    while (
      current &&
      !(current.row === start.row && current.col === start.col)
    ) {
      path.unshift(current);
      current = parent[current.row][current.col];
    }

    for (const cell of path) {
      if (!(cell.row === end.row && cell.col === end.col)) {
        states[cell.row][cell.col] = GRID_ELEMENT_STATES.PATH;
      }
    }

    steps.push({
      grid: grid.map(row => [...row]),
      states: states.map(row => [...row]),
      description: `Path found! Length: ${path.length} cells, Cost: ${gScore[end.row][end.col]}`,
    });
  } else {
    steps.push({
      grid: grid.map(row => [...row]),
      states: states.map(row => [...row]),
      description: 'No path found',
    });
  }

  return steps;
}

/**
 * Pure A* implementation for testing (returns path coordinates)
 * @param {Object} start - Start position {row, col}
 * @param {Object} end - End position {row, col}
 * @param {number} rows - Number of rows
 * @param {number} cols - Number of columns
 * @returns {Object[]|null} - Array of path coordinates or null if no path
 */
export function aStarPure(start, end, rows, cols) {
  const gScore = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(Infinity));
  const fScore = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(Infinity));
  const visited = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(false));
  const parent = Array(rows)
    .fill(null)
    .map(() => Array(cols).fill(null));

  gScore[start.row][start.col] = 0;
  fScore[start.row][start.col] = manhattanDistance(start, end);

  const openSet = [
    {
      row: start.row,
      col: start.col,
      f: fScore[start.row][start.col],
    },
  ];

  const directions = [
    { row: -1, col: 0 },
    { row: 1, col: 0 },
    { row: 0, col: -1 },
    { row: 0, col: 1 },
  ];

  while (openSet.length > 0) {
    openSet.sort((a, b) => a.f - b.f);
    const current = openSet.shift();

    if (visited[current.row][current.col]) {
      continue;
    }

    visited[current.row][current.col] = true;

    if (current.row === end.row && current.col === end.col) {
      const path = [];
      let curr = end;
      while (curr) {
        path.unshift(curr);
        curr = parent[curr.row][curr.col];
      }
      return path;
    }

    for (const dir of directions) {
      const newRow = current.row + dir.row;
      const newCol = current.col + dir.col;

      if (
        newRow >= 0 &&
        newRow < rows &&
        newCol >= 0 &&
        newCol < cols &&
        !visited[newRow][newCol]
      ) {
        const tentativeG = gScore[current.row][current.col] + 1;

        if (tentativeG < gScore[newRow][newCol]) {
          parent[newRow][newCol] = { row: current.row, col: current.col };
          gScore[newRow][newCol] = tentativeG;
          fScore[newRow][newCol] =
            tentativeG + manhattanDistance({ row: newRow, col: newCol }, end);

          openSet.push({
            row: newRow,
            col: newCol,
            f: fScore[newRow][newCol],
          });
        }
      }
    }
  }

  return null;
}
